### ISSUE TITLE
Add support for push notifications

### ISSUE THREAD

**Issue by vlidholt** [serverpod/serverpod#439]
There should be built in libraries for doing push notifications so that everyone won't need to wing their own.

**Comment by II11II** [comment]
https://github.com/invertase/dart_firebase_apis

**Comment by II11II** [comment]
https://github.com/google/googleapis.dart/blob/master/generated/googleapis/lib/fcm/v1.dart

**Comment by p4-k4** [comment]
As with almost every open source BaaS, PN is the missing peice - Putting in a vote for PN here.

**Comment by lukehutch** [comment]
I really need push notifications, and just now realized Serverpod doesn't support them yet...

There's a discussion of tradeoffs of different approaches here:

https://pub.dev/packages/push

But I wonder if Serverpod could skip using Firebase Cloud Messaging entirely, and have its own service in the Flutter client listening to notifications from the Serverpod server...

**Comment by lukehutch** [comment]
FCM push notifications can be sent from the server using the Firebase Admin SDK for FCM:

https://firebase.google.com/docs/cloud-messaging/server#firebase-admin-sdk-for-fcm

But to do this without depending upon the Firebase libraries, the FCM REST API can be called directly:

https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages

--

This is only orthogonally related, but may get at deeper push notification issues, so I throw this out here too:

I have been looking at adding push notifications to the chat server, but I have come across two issues:

1. The Redis pub/sub model only tracks subscriptions for a single server, and there may be different users subscribed to different servers;
2. The current chat server API requires the user's server code to decide whether a user can join a given chat channel or not. The chat channels that any given userId should be able to join are not stored anywhere that the Serverpod chat module knows anything about. That means that there is no way to know what users are part of a chat channel, but are not currently subscribed to the stream of updates from that channel (which is the situation where a push notification should be sent out). Furthermore, the mapping from channel to users that are part of the channel should persist across server restarts, and REDIS doesn't do that. 

So basically, more infrastructure needs to be built in Serverpod to track notification subscriptions, and user liveness.

**Comment by vlidholt** [comment]
@lukehutch You can configure the chat module to use Redis by setting `postMessagesGlobally` to true.
https://github.com/serverpod/serverpod/blob/main/modules/serverpod_chat/serverpod_chat_server/lib/src/business/config.dart

You currently need to keep track of the channels yourself, that is not automatically covered by the chat API.

**Comment by lukehutch** [comment]
Right, I have that option set already. However something missing from the current system is a presence-tracking system (who is currently connected via a socket?). If that were in place, I could find the difference between who is supposed to be able to access a channel (in my own code) and who is currently connected (via the presence system), in order to figure out whether to push a message just to the channel, or out via the push notification API.

I may extend the chat server to store the currently-connected users in Redis, so that no matter where they are connected, whether a user is active can be queried from any Redis node. Is that a change you'd want to see incorporated into Serverpod?

**Comment by vlidholt** [comment]
@lukehutch, that would be a very cool feature to incorporate! :) Perhaps it should be part of the authentication module over the chat module as it can be very useful outside the chat too. It should definitely be a configurable and opt in though as it may incur costs on Redis and Redis isn't turned on by default.

**Comment by lukehutch** [comment]
I have been looking extensively at how to make a single Serverpod server scale to millions of daily active users.

One of the biggest scalability bottlenecks I have run into is that every client that opens a streaming connection uses up a port on the server, and a single IP address can utilize a max of 65536 ports concurrently.

I know that web sockets are very useful, but they don't give you a highly scalable client/server architecture.

I plan to instead use Firebase messaging to push messages from the server to the client, and the regular HTTPS request-based system to send requests from the client to the server. This should scale much better in the long run.

Figured I should mention this here, just to make the point that scalability is a very big reason for Serverpod to support server-origin cloud messaging out of the box.

**Comment by dufguix** [comment]
> One of the biggest scalability bottlenecks I have run into is that every client that opens a streaming connection uses up a port on the server, and a single IP address can utilize a max of 65536 ports concurrently.

@lukehutch   Where do you see 1 stream connection = 1 server port used ?

**Comment by lukehutch** [comment]
It is my understanding that websockets always use one port per connection.

Actually on a related note, does serverpod-chat multiplex all subscribed chat channels through one websocket connection per connected user?

**Comment by dufguix** [comment]
I currently have no idea how serverpod handles websocket. But in general, websockets are handled through 80 and 443.

I am not an expert on real-time web communication but from my little experience on sysadmin, I've never opened, on a firewall, a range of outgoing ports from websocket servers.

**Comment by lukehutch** [comment]
> I currently have no idea how serverpod handles websocket. But in general, websockets are handled through 80 and 443.

Soery, you're right, the limit is 64k ports *per client per server port", not per server.

https://stackoverflow.com/a/2332756/3950982

However the real limit appears to be the number of available file descriptors.

The conclusion is still about the same -- from the link: "The realistic limits being boasted about for normal boxes are around 80K for example single threaded Jabber messaging servers."

Presumably this is a CPU limit, but even for multithreaded servers, you wouldn't be able to support more than an order of magnitude more concurrent connections than that.

**Comment by SandPod** [comment]
Additional details in this issue: https://github.com/serverpod/serverpod/issues/3524

**Comment by abdulawalarif** [comment]
Hi, I am working on the issue.

**Comment by abdulawalarif** [comment]
I’ve opened a PR that adds first-party push as serverpod_push: device registration, a durable queue, and FCM / OneSignal / SNS. 
It’s functional now; please try it and let me know.
